### ISSUE TITLE
XGBoost hyperparameter tuning

### ISSUE THREAD

**Issue by PiranitaGomez** [KUAS-ubicomp-lab/GlucoPatrol#1]
- Need to expand search range
- Need to tune more parameters

https://www.analyticsvidhya.com/blog/2016/03/complete-guide-parameter-tuning-xgboost-with-codes-python/

**Comment by PiranitaGomez** [comment]
```r 

# R code snippet

nrounds <- 1000
tune1 <- expand.grid(
    nrounds = seq(from=200, to=nrounds, by=50),
    eta = c(0.025, 0.05, 0.1, 0.3),
    max_depth = c(4, 6, 8, 10),
    gamma =0,
    subsample = 1, 
    colsample_bytree = 1,
    min_child_weight = 1
    )

xgb_tune1 <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = tune1,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

tune2 <- expand.grid(
    nrounds = seq(from = 100, to = nrounds, by=50),
    eta = xgb_tune1$bestTune$eta,
    max_depth=ifelse(xgb_tune1$bestTune$max_depth == 2, 
        c(xgb_tune1$bestTune$max_depth:4), 
        xgb_tune1$bestTune$max_depth - 1:xgb_tune1$bestTune$max_depth + 1),
    gamma = 0,
    colsample_bytree = 1,
    min_child_weight = c(1,2,3),
    subsample = 1
    )

xgb_tune2 <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = tune2,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

tune3 <- expand.grid(
    nrounds = seq(from = 100, to = nrounds, by=50),
    eta = xgb_tune1$bestTune$eta,
    max_depth = xgb_tune2$bestTune$max_depth,
    gamma = 0,
    colsample_bytree = c(0.4, 0.6, 0.8, 1.0),
    min_child_weight = xgb_tune2$bestTune$min_child_weight,
    subsample = c(0.5, 0.75, 1.0)
    )

xgb_tune3 <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = tune3,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

tune4 <- expand.grid(
  nrounds = seq(from = 100, to = nrounds, by = 50),
  eta = xgb_tune$bestTune$eta,
  max_depth = xgb_tune2$bestTune$max_depth,
  gamma = c(0, 0.05, 0.1, 0.5, 0.7, 0.9, 1.0),
  colsample_bytree = xgb_tune3$bestTune$colsample_bytree,
  min_child_weight = xgb_tune2$bestTune$min_child_weight,
  subsample = xgb_tune3$bestTune$subsample
)

xgb_tune4 <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = tune4,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

tune5 <- expand.grid(
  nrounds = seq(from = 100, to = 10000, by = 100),
  eta = c(0.01, 0.015, 0.025, 0.05, 0.1),
  max_depth = xgb_tune2$bestTune$max_depth,
  gamma = xgb_tune4$bestTune$gamma,
  colsample_bytree = xgb_tune3$bestTune$colsample_bytree,
  min_child_weight = xgb_tune2$bestTune$min_child_weight,
  subsample = xgb_tune3$bestTune$subsample
)

xgb_tune5 <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = tune5,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

final.grid <- expand.grid(
  nrounds = xgb_tune5$bestTune$nrounds,
  eta = xgb_tune5$bestTune$eta,
  max_depth = xgb_tune5$bestTune$max_depth,
  gamma = xgb_tune5$bestTune$gamma,
  colsample_bytree = xgb_tune5$bestTune$colsample_bytree,
  min_child_weight = xgb_tune5$bestTune$min_child_weight,
  subsample = xgb_tune5$bestTune$subsample
)

tune.model <- caret::train(label ~ ., data = ProcessedTrain,
    method = "xgbTree",
    trControl = trctrl,
    #na.action = na.pass,
    tuneGrid = final.grid,
    eval_metric = "mlogloss",
    num_class = 4,
    verbose = TRUE
    )

pred <- predict(tune.model, ProcessedTest)
return(pred)

 ``` </pre>

**Comment by PiranitaGomez** [comment]
# Deeper into Hyperparameter Tuning

### Learning rate
> “The learning rate is the most important hyperparameter in machine learning.” 
> — Andrew Ng

![Image](https://github.com/user-attachments/assets/b05fba73-e37d-434b-8de2-f6d0b9a1adf3)
 
- 0.001-0.01 is often a good starting point.
- 0.01-0.3 is typical range.
- Learning rate of 0.1 is often a good choice.
- Learning rate smaller than 0.001 may take longer time to train without actual improvement in model performance.

[Tune Learning Rate for Gradient Boosting with XGBoost in Python](https://machinelearningmastery.com/tune-learning-rate-for-gradient-boosting-with-xgboost-in-python/)
#The most credible blog about ML on the Internet, by Dr. Jason Brownlee

### Typical ranges for other hyperparameters
[Understanding Hyperparameters in Decision Trees: XGBoost and LightGBM](https://medium.com/@amitsinghrajput_92567/understanding-hyperparameters-in-decision-trees-xgboost-and-lightgbm-7b64cfed77f0)
